File: packages/sillytavern-compat/src/execute.ts
import { compileRegex } from "./regex";
import { previewSillyTavernImport } from "./import";
import type { SillyTavernRegexTestResultView, TestSillyTavernRegexInput, TextTransformRuleView } from "./types";

const MAX_INPUT_BYTES = 4 * 1024 * 1024;
const MAX_OUTPUT_BYTES = 8 * 1024 * 1024;

export async function testSillyTavernRegex(input: TestSillyTavernRegexInput): Promise<SillyTavernRegexTestResultView> {
  const preview = await previewSillyTavernImport(input);
  const macros = { ...macrosFromSpec(preview.contextSpec), ...input.macros };
  return applyTextTransforms(input.input, preview.textTransforms, {
    target: input.target, surface: input.surface, depth: input.depth ?? null,
    isEdit: input.isEdit ?? false, macros,
  });
}

export function applyTextTransforms(
  input: string,
  rules: TextTransformRuleView[],
  context: { target: string; surface: string; depth: number | null; isEdit: boolean; macros: Record<string, string> },
) {
  if (new TextEncoder().encode(input).byteLength > MAX_INPUT_BYTES) throw new Error("text transform input exceeds four MiB");
  let text = input; const appliedRuleIds: string[] = [];
  for (const rule of rules) {
    if (!applies(rule, context)) continue;
    const next = applyRule(text, rule, context.macros);
    if (next !== text) appliedRuleIds.push(rule.id);
    text = next;
    if (new TextEncoder().encode(text).byteLength > MAX_OUTPUT_BYTES) throw new Error(`output exceeds eight MiB after ${rule.id}`);
  }
  return { text, appliedRuleIds };
}

function applies(rule: TextTransformRuleView, context: { target: string; surface: string; depth: number | null; isEdit: boolean }) {
  if (rule.disabled || context.isEdit && !rule.runOnEdit || !rule.targets.includes(context.target as never) || !rule.surfaces.includes(context.surface as never)) return false;
  if (context.depth === null) return true;
  return (rule.minDepth === null || context.depth >= rule.minDepth) && (rule.maxDepth === null || context.depth <= rule.maxDepth);
}

function applyRule(input: string, rule: TextTransformRuleView, macros: Record<string, string>) {
  const source = rule.patternMacroMode === "none" ? rule.findRegex : substitute(rule.findRegex, macros, rule.patternMacroMode === "escaped");
  const regex = compileRegex(source); const global = regex.global && !regex.sticky;
  let output = ""; let cursor = 0; let matched = false;
  while (true) {
    const match = regex.exec(input); if (!match) break;
    output += input.slice(cursor, match.index) + replacement(rule, match, input, macros);
    cursor = match.index + match[0].length; matched = true;
    if (!global) break;
    if (match[0] === "") regex.lastIndex++;
  }
  return matched ? output + input.slice(cursor) : input;
}

function replacement(rule: TextTransformRuleView, match: RegExpExecArray, input: string, macros: Record<string, string>) {
  const value = rule.replaceString.replace(/\$([$&`']|\d{1,2}|<[^>]+>)/g, (token, reference: string) => {
    if (reference === "$") return "$";
    if (reference === "&") return trimmed(match[0], rule, macros);
    if (reference === "`") return input.slice(0, match.index);
    if (reference === "'") return input.slice(match.index + match[0].length);
    if (reference.startsWith("<")) return trimmed(match.groups?.[reference.slice(1, -1)] ?? "", rule, macros);
    return trimmed(match[Number(reference)] ?? token, rule, macros);
  });
  return substitute(value, macros, false);
}

function trimmed(value: string, rule: TextTransformRuleView, macros: Record<string, string>) {
  return rule.trimStrings.reduce((text, trim) => text.replaceAll(substitute(trim, macros, false), ""), value);
}

function substitute(text: string, macros: Record<string, string>, escaped: boolean) {
  return Object.entries(macros).reduce((result, [name, value]) => result.replaceAll(`{{${name}}}`, escaped ? escapeRegex(value) : value), text);
}

function escapeRegex(value: string) { return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"); }
function macrosFromSpec(spec: import("@zhuangsheng/api-client").JsonObject | null) { const value = spec?.textTransformMacros; return value && !Array.isArray(value) && typeof value === "object" ? value as Record<string, string> : {}; }

File: crates/core/src/llm/text_transform/execute.rs
use fancy_regex::Captures;

use super::pattern::{compile_pattern, substitute_macros};
use super::{TextTransformContext, TextTransformOutput, TextTransformRule};
use crate::llm::{LlmConfigError, LlmConfigResult};

const MAX_INPUT_BYTES: usize = 4 * 1024 * 1024;
const MAX_OUTPUT_BYTES: usize = 8 * 1024 * 1024;

pub fn apply_text_transforms(
    input: &str,
    rules: &[TextTransformRule],
    context: &TextTransformContext,
) -> LlmConfigResult<TextTransformOutput> {
    if input.len() > MAX_INPUT_BYTES {
        return Err(error(
            "text_transform_input_limit",
            "text transform input exceeds four MiB",
        ));
    }
    let mut text = input.to_owned();
    let mut applied_rule_ids = Vec::new();
    for rule in rules.iter().filter(|rule| applies(rule, context)) {
        let next = apply_rule(&text, rule, context)?;
        if next != text {
            applied_rule_ids.push(rule.id.clone());
        }
        text = next;
        if text.len() > MAX_OUTPUT_BYTES {
            return Err(error(
                "text_transform_output_limit",
                format!("output exceeds eight MiB after {}", rule.id),
            ));
        }
    }
    Ok(TextTransformOutput {
        text,
        applied_rule_ids,
    })
}

fn applies(rule: &TextTransformRule, context: &TextTransformContext) -> bool {
    if rule.disabled || (context.is_edit && !rule.run_on_edit) {
        return false;
    }
    if context
        .target
        .is_some_and(|value| !rule.targets.contains(&value))
        || context
            .surface
            .is_some_and(|value| !rule.surfaces.contains(&value))
    {
        return false;
    }
    context.depth.is_none_or(|depth| {
        rule.min_depth
            .is_none_or(|min| min < 0 || depth >= min as u32)
            && rule.max_depth.is_none_or(|max| depth <= max)
    })
}

fn apply_rule(
    input: &str,
    rule: &TextTransformRule,
    context: &TextTransformContext,
) -> LlmConfigResult<String> {
    let compiled = compile_pattern(rule, &context.macros)?;
    let mut output = String::with_capacity(input.len());
    let mut cursor = 0usize;
    let mut matched_any = false;
    for capture in compiled.regex.captures_iter(input) {
        let capture = capture
            .map_err(|cause| error("text_transform_runtime", format!("{}: {cause}", rule.id)))?;
        let matched = capture.get(0).expect("capture zero");
        output.push_str(&input[cursor..matched.start()]);
        output.push_str(&replacement(input, rule, &capture, context));
        cursor = matched.end();
        matched_any = true;
        if !compiled.global {
            break;
        }
    }
    if !matched_any {
        return Ok(input.to_owned());
    }
    output.push_str(&input[cursor..]);
    Ok(output)
}

fn replacement(
    input: &str,
    rule: &TextTransformRule,
    captures: &Captures<'_>,
    context: &TextTransformContext,
) -> String {
    let source = rule.replace_string.clone();
    let matched = captures.get(0).expect("capture zero");
    let mut output = String::with_capacity(source.len());
    let mut index = 0usize;
    while index < source.len() {
        if source.as_bytes()[index] == b'$'
            && let Some((end, value)) = capture_reference(
                &source,
                index,
                input,
                matched.start(),
                matched.end(),
                captures,
            )
        {
            output.push_str(&trim_capture(value, rule, context));
            index = end;
            continue;
        }
        let character = source[index..].chars().next().expect("valid boundary");
        output.push(character);
        index += character.len_utf8();
    }
    substitute_macros(&output, &context.macros, false)
}

fn capture_reference<'a>(
    source: &str,
    start: usize,
    input: &'a str,
    match_start: usize,
    match_end: usize,
    captures: &'a Captures<'a>,
) -> Option<(usize, &'a str)> {
    let tail = &source[start + 1..];
    if let Some(name) = tail
        .strip_prefix('<')
        .and_then(|value| value.split_once('>'))
    {
        let end = start + name.0.len() + 3;
        return Some((
            end,
            captures.name(name.0).map_or("", |value| value.as_str()),
        ));
    }
    let special = tail.as_bytes().first().copied()?;
    match special {
        b'$' => return Some((start + 2, "$")),
        b'&' => return Some((start + 2, &input[match_start..match_end])),
        b'`' => return Some((start + 2, &input[..match_start])),
        b'\'' => return Some((start + 2, &input[match_end..])),
        _ => {}
    }
    let digits = tail
        .bytes()
        .take_while(|byte| byte.is_ascii_digit())
        .take(2)
        .count();
    if digits == 0 {
        return None;
    }
    let end = start + 1 + digits;
    let index = source[start + 1..end].parse::<usize>().ok()?;
    Some((end, captures.get(index).map_or("", |value| value.as_str())))
}

fn trim_capture(value: &str, rule: &TextTransformRule, context: &TextTransformContext) -> String {
    rule.trim_strings
        .iter()
        .fold(value.to_owned(), |text, trim| {
            text.replace(&substitute_macros(trim, &context.macros, false), "")
        })
}

fn error(code: &'static str, message: impl Into<String>) -> LlmConfigError {
    LlmConfigError::new(code, message)
}
